src/main/scala/csp/Solver.scala

```diff
@@ -0,0 +1,98 @@
+// OscaR is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 2.1 of the License, or
+// (at your option) any later version.
+//
+// OscaR is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Lesser General Public License  for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License along with OscaR.
+// If not, see http://www.gnu.org/licenses/lgpl-3.0.en.html
+
+package csp
+
+import combinator._
+import oscar.cbls._
+import oscar.cbls.core.search.Neighborhood
+
+import scala.concurrent.duration.Duration
+
+case class Solver(oscarModel: Model, bandit: String) {
+
+  private val obj: Objective = oscarModel.obj
+
+  private val simpleNeighborhoods = SimpleNeighborhoods(oscarModel)
+
+  def solve(verbosity: Int, display: Boolean, fileName: String, timeout: Int): Unit = {
+
+    val withTimeout = timeout < Int.MaxValue
+
+    val neighList: List[Neighborhood] = List(
+      simpleNeighborhoods.swapMostViolated() exhaust simpleNeighborhoods.wideningFlip(),
+      simpleNeighborhoods.wideningFlip(),
+      simpleNeighborhoods.swap()
+//      simpleNeighborhoods.shuffle()
+    )
+
+    val restartN: Neighborhood = simpleNeighborhoods.shuffle(indices = Some(oscarModel.mostViolatedCars))
+
+    val banditNeighborhood: Neighborhood = bandit.toLowerCase() match {
+      case "bandit" => BanditCombinator(neighList, ???, 0, obj, ???) saveBestAndRestoreOnExhaust obj
+
+      case "banditaftermove" =>
+        BanditCombinator(neighList, ???, 0, obj, ???) saveBestAndRestoreOnExhaust obj
+
+      case "banditrollingaverage" =>
+        BanditCombinator(neighList, ???, 0, obj, ???) saveBestAndRestoreOnExhaust obj
+
+      case "epsilongreedy" => new EpsilonGreedyBandit(neighList)
+
+      case "epsilongreedynew" => new EpsilonGreedyBanditNew(neighList)
+
+      case "ucbnew" => new UCB1(neighList)
+
+      case "bestslopefirst" => bestSlopeFirst(neighList)
+
+      case "bestsslopefirstnew" => new BestSlopeFirstNew(neighList)
+
+      case "random" => new RandomCombinator(neighList)
+
+      case _ =>
+        println("warning: invalid bandit specified. Defaulting to bestSlopeFirst")
+        bestSlopeFirst(neighList)
+    }
+
+    var search = {
+      banditNeighborhood match {
+        case BanditCombinator(_, _, _, _, _, _, _, _) => banditNeighborhood
+        case _ =>
+          banditNeighborhood onExhaustRestartAfter (restartN, 0, obj, minRestarts =
```

Comment:
Why hardcoding restartsWithoutImprovement value to 0 ?
Rather than setting it to zero and forcing at least 15 restarts, I would suggest to say min 5 restarts  (or lower) and like 3 restarts without improvement.

From the doc : 
`every time the search neighborhood is exhausted, it checks if the search delivered an improvement over the objective function, and the restart is only performed if it could find an improvement at least once in the last "maxRestartWithoutImprovement" descents.`

So in our case a restart will happens at least 5 times and as long as their is at least one improvement in a 3 restart sequence.

Reply:
Forgot to mention yesterday, currently hardcoded values might later be parametrized, so the original values I committed don't have a lot of meaning - it was mostly to let it compile :)